jumeg_causality/jumeg_causality.py

```diff
@@ -0,0 +1,627 @@
+       min_subject: the subject for common brain
+    '''
+    path_list = get_files_from_list(stcs_path)
+    # loop across all filenames
+    for stcs_path in path_list:
+        subjects_dir = os.environ['SUBJECTS_DIR']
+        minpath = subjects_dir + '/%s' % (min_subject)
+        caupath = stcs_path[:stcs_path.rfind('/%s')]
+        fn_stcs_labels = caupath + '/%s_labels_ts.npy' % (condition)
+        _, _, files = os.walk(stcs_path).next()
+        trials = len(files) / 2
+        srcpath = minpath + '/bem/fsaverage-ico-4-src.fif'
+        src_inv = mne.read_source_spaces(srcpath)
+        # Get unfiltered and morphed stcs
+        stcs = []
+        i = 0
+        while i < trials:
+            fn_stc = stcs_path + 'trial%s_fsaverage' % (str(i))
+            stc = mne.read_source_estimate(fn_stc + '-lh.stc',
+                                           subject=min_subject)
+            stcs.append(stc)
+            i = i + 1
+        # Get common labels
+        list_file = minpath + '/Group_ROIs/common/label_list.txt'
+        with open(list_file, 'r') as fl:
+                  file_list = [line.rstrip('\n') for line in fl]
+        fl.close()
+        rois = []
+        labels = []
+        for f in file_list:
+            label = mne.read_label(f)
+            labels.append(label)
+            rois.append(label.name)
+        # Extract stcs in common labels
+        label_ts = mne.extract_label_time_course(stcs, labels, src_inv,
+                                                 mode='pca_flip',
+                                                 return_generator=False)
+        #make label_ts's shape as (sources, samples, trials)
+        label_ts = np.asarray(label_ts).transpose(1, 2, 0)
+        np.save(fn_stcs_labels, label_ts)
+
+def _mvdetrend(data):
+    '''Multivariate polynomial detrend of time series data.
+       refer:http://users.sussex.ac.uk/~lionelb/MVGC/html/mvdetrend.html.
+       Parameter
+       ---------
+       data: narray
+            multi-trial time series data, shape as (variables, obsevations,
+             trials).
+       Return
+       ------
+       Y: detrended time series.
+    '''
+    n, m, N = data.shape
+    pdeg = 1
+    pdeg = pdeg * np.ones(n)
+    x = np.arange(m) + 1
+    mu = np.mean(x)
+    sig = np.std(x, ddof=1)
+    x = (x - mu) / sig
+    P = np.zeros((n, m))
+    p = []
+    for i in xrange(n):
+        d = pdeg[i]
+        d1 = int(d + 1)
+        V = np.zeros((d1, m))
+        V[d1 - 1, :] = np.ones(m)
+        j = int(d)
+        while j > 0:
+            V[j - 1, :] = x * V[j, :]
+            j = j - 1
+        temp = np.mean(data[i, :, :], axis=-1)
+        p.append(np.linalg.lstsq(V.T, temp)[0].T)
+        P[i, :] = p[i][0] * np.ones((1, m))
+        for j in range(1, d1):
+            P[i, :] = x * P[i, :] + p[i][j]
+    Y = np.zeros(data.shape)
+    for r in xrange(N):
+        Y[:, :, r] = data[:, :, r] - P
+    return Y
+
+def normalize_data(fn_ts, factor=1):
```

Comment:
Hi Praveen,

These functions are mainly translated from matlab, I gave the links in
these scripts.

Best wishes,
Qunxi Dong

2015-08-25 16:25 GMT+02:00 Praveen Sripad <notifications@github.com>:

> In jumeg_causality/jumeg_causality.py
> <https://github.com/jdammers/jumeg/pull/94#discussion_r37870975>:
>
> > +        V[d1 - 1, :] = np.ones(m)
> > +        j = int(d)
> > +        while j > 0:
> > +            V[j - 1, :] = x * V[j, :]
> > +            j = j - 1
> > +        temp = np.mean(data[i, :, :], axis=-1)
> > +        p.append(np.linalg.lstsq(V.T, temp)[0].T)
> > +        P[i, :] = p[i][0] * np.ones((1, m))
> > +        for j in range(1, d1):
> > +            P[i, :] = x * P[i, :] + p[i][j]
> > +    Y = np.zeros(data.shape)
> > +    for r in xrange(N):
> > +        Y[:, :, r] = data[:, :, r] - P
> > +    return Y
> > +
> > +def normalize_data(fn_ts, factor=1):
>
> @dongqunxi <https://github.com/dongqunxi> do these functions have a
> reference? e.g. from Scot?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jdammers/jumeg/pull/94/files#r37870975>.
>